tools/ramdump/dumpParser.py

```diff
@@ -0,0 +1,700 @@
+				return None
+
+			premid = mid
+
+		if symbol_size == 0:
+			return (self.symbol_lookup_table[mid][1], addr - self.symbol_lookup_table[mid][0])
+		else:
+			return (self.symbol_lookup_table[mid][1], self.symbol_lookup_table[mid + 1][0] - self.symbol_lookup_table[mid][0])
+
+
+	# Function to unwind the FRAME with generic method of using Frame pointer
+	def unwind_frame_using_framepointer(self, frame, debug=False):
+		high = 0
+		fp = frame.fp
+
+		low = (frame.sp)		# Get the Stack pointer
+		mask = (self.stacksize) - 1	# Prepare Mask for stacksize for ex.stack:0x8000, mask=0x7FFF
+
+		high = (low + mask) & (~mask)   # Get the end of stack with alignment.ALIGN(low, THREAD_SIZE)
+
+		# Verify the Addresses passed
+		if debug:
+			print "fp:{0:x}  sp low:{1:x} high:{2:x}".format(frame.fp, frame.sp, high)
+
+		# Check current frame pointer is within bounds of Stack*/
+		if (fp < (low + 12) or fp + 4 >= high):
+			print "End of stack frame"
+			return -1
+
+		# If the Ramdump is given, Read contents of address from the RAMDUMP file
+		# by using read_word API
+		if self.log_file is None:
+			fp_is_at = self.read_word(frame.fp - 12)
+			sp_is_at = self.read_word(frame.fp - 8)
+			pc_is_at = self.read_word(frame.fp - 4)
+		else:
+			# If only log file is given, Read contents of addr from the populated stack_table
+			fp_is_at = self.stack_lookup(frame.fp - 12)
+			sp_is_at = self.stack_lookup(frame.fp - 8)
+			pc_is_at = self.stack_lookup(frame.fp - 4)
+
+		frame.fp = fp_is_at
+		frame.sp = sp_is_at
+		frame.pc = pc_is_at
+		return 0
+
+	# Function to walk the stack frame when framepointer is enabled
+	def find_stackframe_using_framepointer(self, frame, extra_str='', output_file=None):
+		while True:
+
+			offset = 0
+			# If PC is Null, prefetch abort or Something wrong, Break
+			if frame.pc is None:
+				break
+			if(frame.pc == 0):
+				break
+
+			# Read the Symbol name and offset for a specific PC address from symbol table
+			r = self.symbol_table_lookup(frame.pc)
+			if r is None:
+				symname = 'UNKNOWN'
+				offset = 0x0
+			else:
+				symname, offset = r # Update both symbol name and offset
+
+			pstring = (extra_str + '[<{0:x}>] {1}+0x{2:x}'.format(frame.pc, symname, offset))
+
+			if output_file:
+				out_file.write(pstring + '\n')
+			else:
+				print(pstring)
+
+			# Unwind the complete Stack with the help of Frame pointer register
+			ret = self.unwind_frame_using_framepointer(frame)
+			if ret < 0:
+				break
+
+	# Function to backtrace the stack in Generic method, which uses frame pointer register
+	# First function to get called to unwind
+	def unwind_backtrace_with_framepointer(self, fp, sp, lr, pc, ss):
+		frame = self.Stackframe(fp,sp,lr,pc)
+		frame.fp = fp
+		frame.sp = sp
+		frame.lr = lr
+		frame.pc = pc
+		self.stacksize = ss
+		self.find_stackframe_using_framepointer(frame)
+
+def usage():
+	print '*************************************************************'
+	print '\nUsage: %s [OPTIONS]\n' % sys.argv[0]
```

Comment:
is there no mandatory argument?
Because [ ] means optional in linux, we don't need to put it. Then with only command, parser should work something. Is it right?

Reply:
@sunghan-chang , Script would work without mandatory arguments but it does not display backtrace.  If mandatory arguments are not passed, it displays the usage of "how to use the script ( something similar to ./dumpParser.py -h )

In help command, an example is also given.

Reply:
I saw it. That's why I left this comment. When only "python dumpParser.py" is executed, it will show help message. That means it has mandatory arguments. Let's provide same usages with linux.
```
~$ cp --help
Usage: cp [OPTION]... [-T] SOURCE DEST
  or:  cp [OPTION]... SOURCE... DIRECTORY
  or:  cp [OPTION]... -t DIRECTORY SOURCE...
Copy SOURCE to DEST, or multiple SOURCE(s) to DIRECTORY.
```
```SOURCE``` and ```DEST``` are mandatories. They are not inside [ ].

Reply:
I modified the usage pattern which fits to dumpParse script usage  as below

pradeep@pradeep-ubuntu:~/2017/pradeep_openTizenRT/TizenRT/tools/ramdump$ python dumpParser.py -h
*************************************************************

Usage: dumpParser.py -e ELF_FILE -r DUMP_FILE [OPTIONS]
                  OR
Usage: dumpParser.py -e ELF_FILE -t LOG_FILE [OPTIONS]

Following options are available
	-e, --elf                       tinyara elf file along with path
	-r, --dump_file                 RAM/FLASH dump_file along with path
	-t, --log_file                  Enter Logfile which contains stackdump during assert
	-G, --gdb_path                  Enter gdb tool path
	-N, --nm_path                   Enter nm tool path
	-E, --readelf_path              Enter readelf tool path
	-h, --help                      Show help